guides/source/deploy_performance_tuning.md

```diff
@@ -0,0 +1,146 @@
+
+Threads are less resilient than processes. Certain errors like segmentation faults can destroy the entire process and all threads inside.
+
+Thread-based workers have much faster startup time than process-based workers. This isn't an issue for most long-running Rails servers, but can be important in some cases.
+
+### Hybrid Concurrency
+
+Puma allows forking multiple processes, each of which uses multiple threads. This provides a compromise between process-based and thread-based concurrency. Using multiple threads per process helps memory usage. Multiple processes permit running more Ruby code at the same time since there is one GVL per process.
+
+Hybrid concurrency limits the damage from a segmentation fault or other error that kills a process. A single process dying will kill that process's threads but not threads in other processes.
+
+Choosing Default Settings
+-------------------------
+
+Rails' default settings are chosen for small applications. You can improve performance for your large application that serves a lot of requests by changing settings.
+
+This section contains common sense defaults based on the type and size of your application and the hosts on which it runs. You can improve performance more by testing your application specifically. See "Performance Testing" for details.
+
+If you're using a Ruby implementation without a GVL such as JRuby or TruffleRuby the number of threads and processes will be completely different. These recommendations are only accurate for CRuby. These are also production recommendations. Your development application will have different needs from a production application, and should use a different configuration.
+
+### Threads Per Process
+
+Rails uses 3 threads per process by default. A well-optimized I/O-heavy Rails application should specify 5 or 6 threads per process at maximum. Discourse, for example, benefits from about 5 threads per process. Small applications with less I/O benefit from around 3 threads per process.
+
+From the default Puma configuration:
+
+> As a rule of thumb, increasing the number of threads will increase how much
+> traffic a given process can handle (throughput), but due to CRuby's
+> Global VM Lock (GVL) it has diminishing returns and will degrade the
+> response time (latency) of the application.
+
+To set the number of threads, you can change the call to the +threads+ method in +config/puma.rb+. Or you can set the +RAILS_MAX_THREADS+ environment variable, which will do the same.
+
+### Number of Processes
+
+[Puma's deployment documentation](https://github.com/puma/puma/blob/master/docs/deployment.md) recommends running at least 1.5 processes per available processor core. Automatic methods to determine the number of cores are unreliable. You should specify the number of processes manually.
```

Comment:
Note that 1.5 per core is if you run in pure process mode. If you run threads on top, then it's 1 per core.

I can't find where but I mentioned that to Nate and he said he would update that doc.

Reply:
Sounds good. That makes more sense to me, but I was worried I was missing something. Fixed. Thanks!